torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -1593,10 +1746,10 @@ def _init_param_attributes(self, p: FlatParameter) -> None:
                 appropriate size and then has its storage freed. This will be
                 resized in place and only materialized (via all-gather) as needed.
         Another attribute is set by :func:`_register_post_backward_hooks`:
-            ``_shard_bwd_hook``: it holds the parameter's AccumulateGrad object
+            ``_post_backward_hook_state``: it holds the parameter's AccumulateGrad object
                 and the registered post hook handle.
         """
-        assert hasattr(p, "_is_sharded"), "Parameters should have been sharded during construction."
+        p = handle.flat_param
```

Comment:
When I cleaned up `_lazy_init()`, I made it so that `_init_param_attributes()` is only called once, namely when `_lazy_init()` is not a no-op. In that case, the check `if hasattr(p, "_local_shard"):` is now kind of meaningless.

The only way to `_lazy_init()` not a no-op again is to call `_reset_lazy_init()`, which will call `del p._local_shard` so that `hasattr(p, "_local_shard")` is `False`. In other words, there should be no way that `if hasattr(p, "_local_shard")` is `True` anymore in `_init_param_attributes()`.

Reply:
had second though, it should be OK to not repointing p._local_shard to the p.data if user mutates it. FSDP currently manages p._local_shard completely and use p._local_shard as the source of sharded parameters, basically FSDP does not care or support users to mutate parameters outside of FSDP, as the mutated parameters are not passed to p._local_shard after FSDP construction, maybe we should document somewhere..